### DISCUSSION TITLE
How do I center text? Is the .align() function documented somewhere?

### DISCUSSION THREAD

**Discussion by arrowtype** [coldtype/coldtype#116] [Q&A] [answered]
I'm trying out Coldtype for an animation or two in a blog post about a font I'm making, and finally starting to wrap my head around the syntax and all these shortcuts. It's a very fun system! It does feel somewhat challenging to go from the bits I know into a fuller layout and animation, though...

At the moment, I'm trying to do two things:
1. Center-align two lines of text 
2. Maybe animate them slightly differently3. 

So far, I can get two lines of text ... but they are overlapping. I'm having trouble figuring out the `.align()` function. Is it described somewhere?

<img width="1072" alt="image" src="https://user-images.githubusercontent.com/45946693/182918326-7a550485-9836-4e95-bb19-b00903db3efd.png">

```py
from coldtype import *

fnt = Font.Cacheable("/Users/stephennixon/code/coldtype-test/assets/ShantellSans[BNCE,IRGL,TRAK,ital,wght].ttf")

@animation((1920, 540), timeline=Timeline(80), fmt="png")
def to_svg(f):
    background = P(f.a.r.inset(0)).f(hsl(240,0,.94)) #hsl(240,0,.94)
    return (
        background,
        StSt("TYPOGRAPHY", fnt, 225
            , TRAK=0.125
            , wght=f.e("eio", 1, rng=(0, 1))
            , IRGL=f.e("eeio")
            , BNCE=f.e("l", 2, rng=(.25, 0.75)))
            .align(f.a.r)
            .f(0),
        StSt("IS KINETIC", fnt, 225
            , TRAK=0.125
            , wght=f.e("eio", 1, rng=(1, 0))
            , IRGL=f.e("eeio")
            , BNCE=f.e("l", 2, rng=(.25, 0.75)))
            .align(f.a.r)
            .f(0)),

```

Sorry if this is a very basic question - I'm still trying to figure out how to figure out things. Thanks for any advice!

**Comment by arrowtype** [comment]
Ah, I see that one option is to use `.translate(x, y)` to adjust lines up and down. So, here's what I've landed on for now:

```py
from coldtype import *

fnt = Font.Cacheable("/Users/stephennixon/code/coldtype-test/assets/ShantellSans[BNCE,IRGL,TRAK,ital,wght].ttf")

@animation((1920, 540), timeline=Timeline(40), fmt="png")
def to_svg(f):
    background = P(f.a.r.inset(0)).f(hsl(240,0,.94)) #hsl(240,0,.94)
    return (
        background,
        StSt("TYPOGRAPHY", fnt, 225
            , TRAK=0.125
            , wght=f.e("eio", 1, rng=(0, 1))
            , IRGL=f.e("eeio")
            , BNCE=f.e("l", 2, rng=(.25, 0.75)))
            .align(f.a.r)
            .translate(0, 120)
            .f(0),
        StSt("IS KINETIC", fnt, 225
            , TRAK=0.125
            , wght=f.e("eio", 1, rng=(1, 0))
            , IRGL=f.e("eeio")
            , BNCE=f.e("l", 2, rng=(0.75, .25)))
            .align(f.a.r)
            .translate(0, -120)
            .f(0)),
```

and that looks like this:

https://user-images.githubusercontent.com/45946693/182920021-c6fd1335-a524-42f8-b913-f9ae385b06a2.mov

If this can be directly in the `.align` line, though – or if there's another, better way, I'd still love to learn about it!

**Comment by stenson** [comment, marked as answer]
Good question! `xalign` might be the function you're looking for — definitely in need of some documentation, though used quite a bit particularly in the `examples/animations` folder.

Both `StSt` and `Glyphwise` can handle multi-line text, and in the `Glyphwise` case, you can look at the `.l` property of the `GlyphwiseGlyph` class that gets passed to the callback — i.e. the `x` passed to the `styler` function below. If you print the `x`, you'll see all the properties available to look at when deciding how to style a given glyph.

Also you can "stack" shapes in Coldtype, as in the second example below, i.e. `stack` and then `lead` to get spacing defined.

```python
from coldtype import *

fnt = Font.MutatorSans()

@animation((1920, 540), timeline=Timeline(40), fmt="png")
def with_glyphwise(f):
    def styler(x):
        if x.l == 0:
            wght = f.e("qeio", 1, rng=(0, 1))
        else:
            wght = f.e("seio", 1, rng=(1, 0))
        
        return Style(fnt, 225, wght=wght)

    return (Glyphwise("TYPOGRAPHY\nIS KINETIC", styler)
        .xalign(f.a.r)
        .lead(30)
        .align(f.a.r)
        .f(0))

@animation((1920, 540), timeline=Timeline(40), fmt="png")
def with_stst(f):
    return (P(
            StSt("TYPOGRAPHY", fnt, 225
                , wght=f.e("eio", 1, rng=(0, 1))),
            StSt("IS KINETIC", fnt, 225
                , wght=f.e("eio", 1, rng=(1, 0))))
        .stack()
        .lead(30)
        .xalign(f.a.r)
        .align(f.a.r)
        .f(0))
```

That said, I think the `translate` answer you provided also works very well! I just think the idiom I've enjoy most is: create the shapes, then xalign them, then stack and lead, then align the union of those shapes, as a general strategy.

**Comment by arrowtype** [comment]
This is so helpful – thank you! I got things working in a way I really like, using the double StSt approach.

**Reply by stenson** [reply]
great to hear — also looking forward to that font being released! looks awesome

**Reply by arrowtype** [reply]
Thanks for the kind words! I'm looking forward to releasing it, too. It always takes so much longer than expected to actually finish a typeface. 😅 But also, I am excited to see what people might do with it! This one will be OFL, and is meant to be a bit of a Comic Sans competitor.
